File: sub-web-api/src/main/java/org/github/web/controller/ExceptionController.java
package org.github.web.controller;

import lombok.extern.slf4j.Slf4j;
import org.github.ops.LoggerOpsKt;
import org.github.spring.restful.Returnable;
import org.github.spring.restful.json.JSONReturn;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseBody;

@ControllerAdvice
@Slf4j
public class ExceptionController {
  @ExceptionHandler(IllegalArgumentException.class)
  @ResponseBody
  public Returnable illegalArgumentExceptionHandler(Exception ex) {
    LoggerOpsKt.error(log, ex, ex::getMessage);
    return JSONReturn.error().withRetMsg(ex.getMessage());
  }
}

File: sub-web-api/src/main/java/org/github/web/controller/ExceptionController.kt
package org.github.web.controller

import org.github.exception.ParamsErrorException
import org.github.ops.error
import org.github.ops.log
import org.github.spring.restful.json.JSONReturn.error
import org.github.spring.restful.json.JSONReturn.warn
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice

@RestControllerAdvice
class ExceptionController {
  private val log = ExceptionController::class.log

  @ExceptionHandler(ParamsErrorException::class)
  fun handleParamsErrorException(e: ParamsErrorException) = warn().withRetMsg(e.message.toString()).apply { log.error(e) { e.message } }

  @ExceptionHandler(Throwable::class)
  fun handleThrowable(e: Throwable) = error().withRetMsg(e.message.toString()).apply { log.error(e) { e.message } }
}
